### PR TITLE
chore: make PyAutoGUI default

### LINKED ISSUE #173: make pyautogui default (as it works mss doens't work for everyone)
and write (fast) in brackets in front of d3dshot

### PR DESCRIPTION
**Pull request by Femi-lawal**
Fixed #173

### CODE REVIEW

File: fishy/engine/common/screenshot.py [hunk 1]
```diff
@@ -94,11 +94,25 @@ def setup(self) -> bool:
...
+
+    # Default to PyAutoGUI if D3DShot is not explicitly chosen or fails to set up
+    for lib in LIBS:
+        lib_instance = lib()
+        if lib_instance.setup():
+            logging.debug(f"Using {lib.__name__} as the screenshot library.")
+            return lib_instance
+
+    logging.error("No suitable screenshot library found. Please check your configuration.")
+    raise Exception("Failed to initialize a screenshot library.")
```
Comment (adsau59) [inline fishy/engine/common/screenshot.py:118]: not a good idea to throw exception, might crash the thread if its not handled, logging and error, and returning None, and handling the None value where this method is used is a much cleaner way of handling errors

File: fishy/engine/common/screenshot.py [hunk 2]
```diff
@@ -94,11 +94,25 @@ def setup(self) -> bool:
...
+            return d3dshot_instance
+        else:
+            logging.warning("D3DShot setup failed. Falling back to default method...")
+
+    # Default to PyAutoGUI if D3DShot is not explicitly chosen or fails to set up
+    for lib in LIBS:
+        lib_instance = lib()
+        if lib_instance.setup():
+            logging.debug(f"Using {lib.__name__} as the screenshot library.")
+            return lib_instance
```
Comment (adsau59) [inline fishy/engine/common/screenshot.py:115]: I like the idea of automatically test each lib and use the one which works...

some points to consider...

1. setup() throws exceptions as the screenshot lib being used is unpredictable on some system, so maybe, handle it with try/catch too
2. preferred_lib isn't even used if its not d3dshot, this logic doesn't let user use the library they selected

File: fishy/engine/common/screenshot.py [hunk 3]
```diff
@@ -94,11 +94,25 @@ def setup(self) -> bool:
...
-    return lib()
+    # Attempt to use D3DShot if performance is critical and it's specified in the config
+    preferred_lib = config.get("sslib", None)
+    if preferred_lib == "d3dshot":
+        logging.debug("Attempting to use D3DShot (fast) as the screenshot library...")
+        d3dshot_instance = D3DShot()
+        if d3dshot_instance.setup():
+            return d3dshot_instance
+        else:
+            logging.warning("D3DShot setup failed. Falling back to default method...")
```
Comment (adsau59) [inline fishy/engine/common/screenshot.py:108]: I don't get the reason behind handling d3dshot separately

**Review by adsau59 — CHANGES_REQUESTED** [review]
changes requested

**Comment by Femi-lawal** [comment]
@adsau59 Thank you for the review, I've implemented the changes.

**Review by adsau59 — CHANGES_REQUESTED** [review]
code looks good to me, just need to handle None [here](https://github.com/fishyboteso/fishyboteso/blob/0fdb285aea0d912454699785a8cae6c63d796b95/fishy/engine/common/window_server.py#L38)

**Comment by Femi-lawal** [comment]
> code looks good to me, just need to handle None [here](https://github.com/fishyboteso/fishyboteso/blob/0fdb285aea0d912454699785a8cae6c63d796b95/fishy/engine/common/window_server.py#L38)

@adsau59 Hi, I've updated screenshot.create() to handle None

**Comment by adsau59** [comment]
looks good, thanks for the contribution

**Pull request merged**